src/pp/if_parser.rs

```diff
@@ -0,0 +1,441 @@
+use crate::token::{Integer, PreprocessorError, Punct};
+
+use super::{Define, Location, Step, StepExit, Token, TokenValue};
+use std::{collections::HashMap, iter::Peekable, rc::Rc, slice::Iter, vec::IntoIter};
+
+pub(super) struct IfParser<'macros> {
+    tokens: Peekable<IntoIter<Token>>,
+    defines: &'macros HashMap<String, Rc<Define>>,
```

Comment:
There is a lot of manual handling of macros in this file. MacroExpander in pp.rs does that already and handles function-like macros and everything. `evaluate_expression` could build one with a local `MELexer` a bit like `ExpandParameterLexer` in `pp.rs`. You'd still need the parent lexer to ask it if a define exists or not (and to implement some of `MELexer`).

Note that `defined` is a disallowed define name (although we don't implement that yet), so there is no risk that the `MacroExpander` starts eating `defined` and turning it into something else.